rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,444 @@
+
+### Testing TensorFlow Lite Model correctness
+
+On-device training is complex and it can be hard to troubleshoot. TensorFlow
+Lite should provide guidelines and tools to make this easy. The core idea is: If
+each TensorFlow function is converted to a TensorFlow Lite subgraph without
+changing semantics, it will be easy to test TensorFlow and TensorFlow Lite
+functionality side by side.
+
+In addition to the common ML testing best practices, there are a few specific
+suggestions for TensorFlow Lite on-device learning:
+
+*   **Test all functions in TensorFlow** before converting to TensorFlow Lite.
+    In this use case, in addition to testing `inference` and `train`, also test
+    if `train_dense_layers` since it's used on device. If it doesn't work well
+    in TensorFlow, it can't possibly work well in TensorFlow Lite. Check if
+    there are any coding error, or you may need to tune the model structure or
+    hyperparameter.
+*   **Use model diff testing tool to test all functions**. TensorFlow Lite now
+    has a tool which feeds the same random data into TensorFlow and TensorFlow
+    Lite, and compare if the results are close enough in a threshold. The model
+    should be extended to handle multiple subgraphs (including training and
+    inference subgraphs) and variables. This is an end to end test which can
+    capture kernel bugs or TensorFlow/TensorFlow Lite discrepancy.
+
+## Implementation details for basic training
+
+There is nothing magical in training. Technically, training is doing a series of
+**mathematical computation** (requiring these op kernels), and update the weight
+**variables** which is **accessible by other (inference) subgraphs**. The
+converter also needs to be changed, to support and optimize training use cases.
+
+### Sharing variables across subgraphs
+
+For the variable requirement of TensorFlow Lite training, the high level goals
+are:
+
+*   Convert TensorFlow Resource Variables to TensorFlow Lite, and preserve the
+    same semantic
+*   Able to share variables between multiple TensorFlow Lite subgraphs
+
+We propose to define variables in TFLite that have similar semantics to resource
+variables in TensorFlow:
+
+*   Use int32 tensor to represent variable ID in TFLite instead of defining a
+    resource type
+*   Only 2 ops are required to make it work: `AssignVariableOp` and
+    `ReadVariableOp`
+*   Other ops can be added into TFLite when necessary
+
+### Gradient and optimizer operations
+
+To train a model, it's required to compute **gradients**, and apply the gradient
+to trainable variables with **optimizers** (e.g. SGD, ADAM...etc). Technically,
+gradients and optimizers are just some mathematical computation and variable
+read/write operations. When constructing training graphs in TensorFlow,
+sometimes a single fused op is used (e.g. `ReluGrad`), and sometimes it produces
+multiple regular ops in unfused form (e.g. `Mul`, `MatMul`, `Add`) to compute
+the gradient of one op.
+
```

Comment:
Is there any optimization plan on the storage of intermediate results when training on-device?

Reply:
Could you elaborate more? Are you referring to the intermediate computation result (e.g. intermediate tensors those are neither inputs nor outputs) in memory? or the trained weights?

Reply:
Right, i am referring to the intermediate computation results in memory.